[Core] Add a new codemod package

oliviertassinari opened the pull request:
[react-codemod](https://github.com/reactjs/react-codemod) was a great source of inspiration for this PR.
Thanks @cpojer for this amazing `jscodeshift` package.

I have been using the `import-path.js` script successfully on my code base.

nathanmarks commented on packages/material-ui-codemod/README.md:
```diff
@@ -0,0 +1,41 @@
+
+- `npm install -g jscodeshift`
+- `npm install material-ui-codemod`
+- `jscodeshift -t <codemod-script> <file>`
```
Signature according to https://github.com/facebook/jscodeshift#usage-cli is actually `<path>`, you can do whole directories not just individual files, which `<file>` might lead one to think be the case.

oliviertassinari replied:
That sounds like a good idea to change to `<path>`.
Let's see what facebook think about it https://github.com/reactjs/react-codemod/pull/37 :grin:.

nathanmarks replied:
Haha, nice :+1:

mbrookes commented:
> mbrookes changed the title from [Core] Add a new codemod package to [Core] Add a new package for codemods an hour ago

Race conditions!

nathanmarks commented:
This looks awesome. I gotta check out this `jscodeshift` project... I've been doing this sort of stuff in a much less sophisticated fashion!

oliviertassinari commented:
I have added the support for imports using the `lib` folder.

:warning: The mapping isn't complete. I hope that people will add new entries in the `pathConversion` object is they need to :tada:.

mbrookes commented on packages/material-ui-codemod/README.md:
```diff
@@ -0,0 +1,43 @@
+-import FlatButton from 'material-ui/src/flat-button';
+-import RaisedButton from 'material-ui/lib/raised-button';
++import FlatButton from 'material-ui/src/FlatButton';
++import RaisedButton from 'material-ui/lib/RaisedButton';
```
`import RaisedButton from 'material-ui/RaisedButton';`?

oliviertassinari replied:
That's not true yet. Should I do an optimistic update?

mbrookes replied:
I think so , yes. We should publish the next release as we intend the import paths to be in 0.15.0, so the first time any devs use this codemod, it will be for flattened imports.

oliviertassinari replied:
:+1:

mbrookes commented on packages/material-ui-codemod/transforms/v0.15.0/import-path.js:
```diff
@@ -0,0 +1,69 @@
+};
+
+const pathBaseSource = 'material-ui/src/';
+const pathBasePackage = 'material-ui/lib/';
```
`const pathBasePackage = 'material-ui/';`?

oliviertassinari replied:
Non, that variable is for the existing code :grin:. But I get what you mean. As we plan do remove the `lib` from the require, we should reflect it in the codemod.

mbrookes replied:
:+1: I wasn't too sure. :smile:

oliviertassinari replied:
done;

mbrookes commented on packages/material-ui-codemod/package.json:
```diff
@@ -0,0 +1,7 @@
+  "description": "Codemod scripts for Material-UI",
+  "scripts": {},
+  "license": "MIT"
+}
```
I think npm will warn about missing keys when you publish. `repository` and `license` are mandatory I think, and possibly `author`. Some material-ui related `keywords` might be a good idea to help discoverability.

oliviertassinari replied:
That's a very good feedback. I have rushed this file.

mbrookes commented on packages/material-ui-codemod/README.md:
````diff
@@ -0,0 +1,46 @@
+```
+
+```sh
+jscodeshift -t material-ui-codemod/transforms/v0.15.0/import-path.js <path>
````
`node_modules/material-ui-codemod...`?

oliviertassinari replied:
Well, I guess yes.

mbrookes commented:
It doesn't run for me - even the error has an error! :smile: 

```
$ jscodeshift -d ./import-path.js ./docs
/usr/local/lib/node_modules/jscodeshift/dist/Runner.js:156
    process.stdout.error(clc.whiteBright.bgRed('ERROR') + ' Transform file ' + transformFile + ' does not exist \n');
                   ^

TypeError: process.stdout.error is not a function
    at Object.run (/usr/local/lib/node_modules/jscodeshift/dist/Runner.js:156:20)
    at Object.<anonymous> (/usr/local/lib/node_modules/jscodeshift/bin/jscodeshift.sh:78:8)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Function.Module.runMain (module.js:447:10)
    at startup (node.js:140:18)
    at node.js:1001:3
```

but that seems to be a jscodeshift issue.

oliviertassinari commented:
You have forgotten the `-t` option.

mbrookes commented:
Ah, misunderstood, and thought `-d` was instead of `-t`. In any case, `jscodeshift` doesn't handle malformed params very well. :smile:

mbrookes commented:
Tried running this agains the 0.14.4 docs site wihs:
```
jscodeshift -t ./import-path.js ./docs/src/ -d
```
and it only found 3 files, and modified none.

```
Processing 3 files... 
Spawning 3 workers...
Running in dry mode, no files will be written! 
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
All done. 
Results: 
0 errors
3 unmodified
0 skipped
0 ok
Time elapsed: 1.546seconds 
```

There's no recursion option, so thought perhaps it was an file extension issue (it says it defaults to `js`), but running with `-extensions` throws an error (regardless of quotes or not around the extension):

```
$ jscodeshift -t ./import-path.js ./docs/src/ -extensions 'jsx' -d
path.js:8
    throw new TypeError('Path must be a string. Received ' +
    ^

TypeError: Path must be a string. Received true
    at assertPath (path.js:8:11)
    at Object.posix.resolve (path.js:426:5)
    at Object.<anonymous> (/usr/local/lib/node_modules/jscodeshift/bin/jscodeshift.sh:79:58)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Function.Module.runMain (module.js:447:10)
    at startup (node.js:140:18)
    at node.js:1001:3
```
Someone should teach these `jscodemod` numpties how to write command-line utils! Even my throwaway scripts fail gracefully on params errors.

Time to RTFM. :smile:

mbrookes commented:
Looking at the docs / usage, it seems the path should come first:

```jscodeshift <path>... [options]```

But the end result is the same.

oliviertassinari commented:
@mbrookes This could help you https://github.com/zertosh/class-props-codemod/issues/1#issuecomment-177550943.

mbrookes commented:
So `jscodemod` us utterly borked and we're wasting our time here?

oliviertassinari commented:
> and modified none:

You are using the dry option. That's the intended behavior. 

> There's no recursion option

True, I'm using old fashion sh command. For instance, with the ES6 classes transform, I'm running the following command.
```sh
find src \
    \( \
      -path '*/node_modules/*' \
    \) -prune -o -name '*.js' -print \
  | xargs jscodeshift -t ../react-codemod/transforms/class.js
```

mbrookes commented:
> You are using the dry option. That's the intended behavior.

No, no. Without `-d`. "the only thing it changed was the formatting." I hope it wouldn't do that with '-d'!

> There's no recursion option

By that I meant, "so I assume it recurses by default". Why is it spawning workers if it can only process one?

```
Processing 3 files... 
Spawning 3 workers...
Running in dry mode, no files will be written! 
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
```
Processing only the files in a particular directory seems equally bizarre.

Borked, borked, borked.

Maybe there are better alternatives?

nathanmarks commented:
@mbrookes @oliviertassinari We could set this up with our own CLI and use the API programatically

oliviertassinari commented:
> We could set this up with our own CLI and use the API programatically

We could, but neither facebook nor [rackt](https://github.com/reactjs/rackt-codemod/blob/master/README.md) is doing so.We don't have their man power. 
That's kind of a MVP. I would rather follow the community convention and sent time on the actual source code.

TL;DR I won't do it, but anyone is welcomed to add an extra level around jscodeshift to make it no-brainer.

nathanmarks commented:
@oliviertassinari It's a tiny tiny amount of work if it can help fix the main user experience issue here (no globbing). I can do it.  ([here is an example](https://github.com/nathanmarks/cbioportal-api-client/blob/master/lib/cli.js) from a little util of mine showing how easy it is to set up a node app with a user friendly CLI.

However, this is starting to seem a lot more complex than a simple search/replace in files in your code editor. If users can't do that -- the hiccups @mbrookes encountered may also trouble them.

oliviertassinari commented:
@mbrookes Have you tried the following prefix ?
```sh
find src -name '*.js' -print | xargs
```
I hope that we can document this tool correctly to make people migration simpler :+1:.

mbrookes commented:
Hi @oliviertassinari I haven't, bit since it didn't change the paths correctly even for a single file, feeding it multiple files is probably not going to help. :wink:

newoga commented:
Question: Would it be out of scope for the `import-path` codemod to fix `raw-themes` to `baseThemes` as well?

If we have the compromise implemented, _maybe_ we can just kill the raw-themes this release.

oliviertassinari commented:
@newoga That sounds more or less in the scope of this transform.

oliviertassinari commented:
- I have added some tests
- The new tests are automatically executed by the CI tool
- I have added almost all the components
- The README provides a way to use the tool that is supposed to work out of the box

Please review :tada:

nathanmarks commented on .travis.yml:
```diff
@@ -4,5 +4,6 @@ node_js:
 sudo: false
 script:
   - npm run lint
+  - cd packages/material-ui-codemod && npm install && cd ../../
```
Hey @oliviertassinari , I think if this line is required for `npm run test` to execute without error, that those dependencies need to be installed to a `postinstall` script hook.

Otherwise, we should keep the tests for this separate -- similar to how `icon-builder` has it's own tests.

I'm fine with either one, I just don't like the idea of `npm install && npm run test` failing because the user didn't first navigate to a subdirectory and `npm install`.

oliviertassinari replied:
I don't like this idea either. I was looking for the best way to do it. Using the `postinstall` sounds far better. Thanks.

nathanmarks commented on test/unit.js:
```diff
@@ -13,7 +13,23 @@ const mocha = new Mocha({
 
-Glob(`src/**/${argv.component ? argv.component : '*'}.spec.js`, {}, (err, files) => {
+const paths = [
+  `src/**/*${argv.component ? argv.component : ''}*.spec.js`,
```
Can we put the `*` back inside the ternary on this line? I moved it in a PR so that `-c Step` for example wouldn't run tests for every component beginning with "Step" (StepButton, StepLabel, Stepper, StepContent, StepConnector, etc...)

oliviertassinari replied:
Oups, my bad.

nathanmarks replied:
Np! Was a tiny change I made, hard to notice :smile:

oliviertassinari replied:
@nathanmarks Good eyes! :eyes:

nathanmarks commented on test/unit.js:
```diff
@@ -13,7 +13,23 @@ const mocha = new Mocha({
+];
+
+if (!argv.component) {
+  paths.push('packages/material-ui-codemod/transforms/**/*.spec.js');
```
As I mentioned above -- I think if we're running these tests from the main test suite that we need to have a better way of making sure all the dependencies are installed for `npm run test` without the user needing to first go to that package subfolder and running `npm install`.

nathanmarks commented:
Hey @oliviertassinari , looking good for the most part :+1: I left 1 or 2 notes related to the test setup

oliviertassinari commented:
@nathanmarks I have added a `postinstall` script.

nathanmarks commented:
@oliviertassinari quick note RE our chat on gitter:

Let's provide copy/pastable paths for the scripts people need to use next to the instructions:

```
node_modules/material-ui-codemod/transforms/v0.15.0/import-path.js
```

nathanmarks commented on packages/material-ui-codemod/transforms/v0.15.0/import-path.js:
```diff
@@ -0,0 +1,102 @@
+  'card/card-text': 'Card/CardText',
+  'date-picker/date-picker': 'DatePicker',
+  'dialog': 'Dialog',
+  'divier': 'Divider',
```
typo
